## 1.0 Import the necessary libraries

In [1]:
# Code task 1#
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime

## 2.0 Import the Datasets

### 2.1 Import the first dataset

In [2]:
# Code task 2#
# Loading the takehome_user_engagement dataset 
user_engagement = pd.read_csv('takehome_user_engagement.csv')

# Code task 3#
# Display the first few rows to verify successful loading
user_engagement.head()

time_stamp  user_id  visited
0  2014-04-22 03:53:30        1        1
1  2013-11-15 03:45:04        2        1
2  2013-11-29 03:45:04        2        1
3  2013-12-09 03:45:04        2        1
4  2013-12-25 03:45:04        2        1

In [3]:
#Code task 5#
# Check on the column
user_engagement.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207917 entries, 0 to 207916
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   time_stamp  207917 non-null  object
 1   user_id     207917 non-null  int64 
 2   visited     207917 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 4.8+ MB


In [4]:
# Code task 6#
# Convert time_stamp in the appropriate format
user_engagement['time_stamp'] = pd.to_datetime(user_engagement['time_stamp'])

# Code task 7#
# Check on the column after changing
user_engagement.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207917 entries, 0 to 207916
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   time_stamp  207917 non-null  datetime64[ns]
 1   user_id     207917 non-null  int64         
 2   visited     207917 non-null  int64         
dtypes: datetime64[ns](1), int64(2)
memory usage: 4.8 MB


### 2.2 Import the second dataset

In [5]:
# Code task 8#
# Loading the takehome_user encoding (ISO-8859-1)
users = pd.read_csv('takehome_users.csv', encoding='ISO-8859-1')

# Code task 9#
# Display the first few rows to verify successful loading
users.head()

object_id        creation_time               name  \
0          1  2014-04-22 03:53:30     Clausen August   
1          2  2013-11-15 03:45:04      Poole Matthew   
2          3  2013-03-19 23:14:52  Bottrill Mitchell   
3          4  2013-05-21 08:09:28    Clausen Nicklas   
4          5  2013-01-17 10:14:20          Raw Grace   

                        email creation_source  last_session_creation_time  \
0    AugustCClausen@yahoo.com    GUEST_INVITE                1.398139e+09   
1      MatthewPoole@gustr.com      ORG_INVITE                1.396238e+09   
2  MitchellBottrill@gustr.com      ORG_INVITE                1.363735e+09   
3   NicklasSClausen@yahoo.com    GUEST_INVITE                1.369210e+09   
4          GraceRaw@yahoo.com    GUEST_INVITE                1.358850e+09   

   opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                         1                           0      11   
1                         0                           0       1   
2                         0                           0      94   
3                         0                           0       1   
4                         0                           0     193   

   invited_by_user_id  
0             10803.0  
1               316.0  
2              1525.0  
3              5151.0  
4              5240.0

In [6]:
# Code task 10#
# Check on the columns
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   object_id                   12000 non-null  int64  
 1   creation_time               12000 non-null  object 
 2   name                        12000 non-null  object 
 3   email                       12000 non-null  object 
 4   creation_source             12000 non-null  object 
 5   last_session_creation_time  8823 non-null   float64
 6   opted_in_to_mailing_list    12000 non-null  int64  
 7   enabled_for_marketing_drip  12000 non-null  int64  
 8   org_id                      12000 non-null  int64  
 9   invited_by_user_id          6417 non-null   float64
dtypes: float64(2), int64(4), object(4)
memory usage: 937.6+ KB


In [7]:
# Code task 11#
# Convert the column 'creation_time' in the appropriated format
users['creation_time']= pd.to_datetime(users['creation_time'])

# Code task 12#
# Check after the change
users.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   object_id                   12000 non-null  int64         
 1   creation_time               12000 non-null  datetime64[ns]
 2   name                        12000 non-null  object        
 3   email                       12000 non-null  object        
 4   creation_source             12000 non-null  object        
 5   last_session_creation_time  8823 non-null   float64       
 6   opted_in_to_mailing_list    12000 non-null  int64         
 7   enabled_for_marketing_drip  12000 non-null  int64         
 8   org_id                      12000 non-null  int64         
 9   invited_by_user_id          6417 non-null   float64       
dtypes: datetime64[ns](1), float64(2), int64(4), object(3)
memory usage: 937.6+ KB


## 3.0 Identify the adopter user

In [8]:
# Code task 13#
# Extract date from timestamp for unique login tracking
user_engagement["date"] = user_engagement["time_stamp"].dt.date

# Code task 14#
# Sort values for correct rolling window application
user_engagement = user_engagement.sort_values(by=["user_id", "date"])

# Code task 15#
# Function to check if a user has logged in 3 times within a 7-day window
def is_adopted(user_dates):
    user_dates = sorted(user_dates.unique())  # Ensure unique dates and sorted order
    return any((user_dates[i + 2] - user_dates[i]).days <= 7 for i in range(len(user_dates) - 2))

# Code task 16#
# Group by user_id and check for adoption criteria
adopted_users = user_engagement.groupby("user_id")["date"].apply(is_adopted)

# Code task 17#
# Convert adopted users to a DataFrame
adopted_users_df = adopted_users.reset_index()

# Code task 18#
# Filter for adopted users
adopted_users_df = adopted_users_df[adopted_users_df['date'] == True]

# Code task 19#
# Rename the 'date' column to 'adopted'
adopted_users_df.rename(columns={'date': 'adopted'}, inplace=True)

# Code task 20#
# Display the first few rows of the adopted users DataFrame
adopted_users_df.head()


user_id  adopted
1         2     True
7        10     True
13       20     True
24       33     True
28       42     True

## 4.0 Factors predict future user adoption

In [9]:
# Code task 21#
# Merge adopted users data with the users dataset
users = users.rename(columns={"object_id": "user_id"})  # Ensure correct key for merging

# Code task 22#
# Merge to add adoption status to the user data
users_adopted = users.merge(adopted_users_df, on="user_id", how="left")

# Code task 23#
# Fill missing values in 'adopted' column with 0 (indicating non-adopted users)
users_adopted["adopted"] = users_adopted["adopted"].fillna(0).astype(int)

# Code task 24#
# Display the first few rows of the merged dataset
users_adopted.head()

user_id       creation_time               name                       email  \
0        1 2014-04-22 03:53:30     Clausen August    AugustCClausen@yahoo.com   
1        2 2013-11-15 03:45:04      Poole Matthew      MatthewPoole@gustr.com   
2        3 2013-03-19 23:14:52  Bottrill Mitchell  MitchellBottrill@gustr.com   
3        4 2013-05-21 08:09:28    Clausen Nicklas   NicklasSClausen@yahoo.com   
4        5 2013-01-17 10:14:20          Raw Grace          GraceRaw@yahoo.com   

  creation_source  last_session_creation_time  opted_in_to_mailing_list  \
0    GUEST_INVITE                1.398139e+09                         1   
1      ORG_INVITE                1.396238e+09                         0   
2      ORG_INVITE                1.363735e+09                         0   
3    GUEST_INVITE                1.369210e+09                         0   
4    GUEST_INVITE                1.358850e+09                         0   

   enabled_for_marketing_drip  org_id  invited_by_user_id  adopted  
0                           0      11             10803.0        0  
1                           0       1               316.0        1  
2                           0      94              1525.0        0  
3                           0       1              5151.0        0  
4                           0     193              5240.0        0

In [10]:
# Code task 25#
# Convert categorical variables into numerical format for correlation analysis
users_adopted_encoded = users_adopted.copy()
users_adopted_encoded = pd.get_dummies(users_adopted_encoded, columns=["creation_source"], drop_first=True)

# Code task 26#
# Select only numerical columns for correlation analysis
numeric_columns = users_adopted_encoded.select_dtypes(include=[np.number])

# Code task 27#
# Calculate the correlation matrix
correlation_matrix = numeric_columns.corr()

# Code task 28#
# Extract correlations with the 'adopted' column
correlation_with_adoption = correlation_matrix["adopted"].sort_values(ascending=False)

# Code task 29#
# Display the correlations
print(correlation_with_adoption)

adopted                       1.000000
last_session_creation_time    0.392369
org_id                        0.066995
opted_in_to_mailing_list      0.008838
enabled_for_marketing_drip    0.006578
user_id                       0.005292
invited_by_user_id            0.001185
Name: adopted, dtype: float64


### 4.1 Summary

The correlation analysis highlights which factors are most related to user adoption:

. Last Session Creation Time (0.39)

    - This is the strongest positive correlation.
    - Users who logged in recently are much more likely to be adopted users.

. Organization ID (0.067)
    - Users in certain organizations are slightly more likely to be adopted.
    - This suggests some organizations have higher engagement levels.
        
. Marketing Opt-in (0.0088) and Marketing Drip (0.0066)
    - Users who opt into marketing emails or are included in marketing drip campaigns show a very weak correlation with adoption.
    - This suggests marketing campaigns alone are not a strong predictor of user adoption.

END